python/plugins/processing/algs/gdal/GdalAlgorithm.py

```diff
@@ -87,8 +87,13 @@ def getOgrCompatibleSource(self, parameter_name, parameters, context, feedback,
             if executing:
                 # parameter is not a vector layer - try to convert to a source compatible with OGR
                 # and extract selection if required
+                if 'gpkg' in QgsVectorFileWriter.supportedFormatExtensions():
+                    preferred_format = 'gpkg'
+                else:
+                    preferred_format = 'shp'
                 ogr_data_path = self.parameterAsCompatibleSourceLayerPath(parameters, parameter_name, context,
                                                                           QgsVectorFileWriter.supportedFormatExtensions(),
+                                                                          preferred_format,
```

Comment:
`QgsVectorFileWriter.supportedFormatExtensions()[0]` will works more simply.

What I got on my system when GDAL is built:

Without external option

```
QgsVectorFileWriter.supportedFormatExtensions()
['shp', '000', 'bna', 'csv', 'dgn', 'dxf', 'geojson', 'geojsonl', 'geojsons', 'gml', 'gpx', 'gxt', 'json', 'kml', 'tab', 'txt', 'xml']
```

With gpkg/sqlite3 (and other formats):

```
QgsVectorFileWriter.supportedFormatExtensions()
['gpkg', 'shp', '000', 'bna', 'csv', 'dgn', 'dxf', 'geojson', 'geojsonl', 'geojsons', 'gml', 'gpx', 'gxt', 'ili', 'itf', 'json', 'kml', 'ods', 'sqlite', 'tab', 'txt', 'xlsx', 'xml', 'xtf']
```

By default, supportedFormatExtensions() returns the list sorted by recommanded formats, so it's not necessary to add this if/else statement.